protocol/persistence_metrics.go

```diff
@@ -0,0 +1,58 @@
+package protocol
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+)
+
+const selectTimestampsQuery = "SELECT whisper_timestamp FROM user_messages WHERE %s whisper_timestamp >= ? AND whisper_timestamp <= ?"
+const selectCountQuery = "SELECT COUNT(*) FROM user_messages WHERE %s whisper_timestamp >= ? AND whisper_timestamp <= ?"
+
+func querySeveralChats(chatIDs []string) string {
+	if len(chatIDs) == 0 {
+		return ""
+	}
+
+	var conditions []string
+	for _, chatID := range chatIDs {
+		conditions = append(conditions, fmt.Sprintf("local_chat_id = '%s'", chatID))
+	}
+	return fmt.Sprintf("(%s) AND", strings.Join(conditions, " OR "))
+}
+
+func (db sqlitePersistence) SelectMessagesTimestampsForChatsByPeriod(chatIDs []string, startTimestamp uint64, endTimestamp uint64) ([]uint64, error) {
+	query := fmt.Sprintf(selectTimestampsQuery, querySeveralChats(chatIDs))
+
+	rows, err := db.db.Query(query, startTimestamp, endTimestamp)
+	if err != nil {
+		return []uint64{}, err
+	}
+	defer rows.Close()
+
+	var timestamps []uint64
+	for rows.Next() {
```

Comment:
The section `Fetching Data from the Database` is quite useful, as is the whole page actually http://go-database-sql.org/retrieving.html. It says we should ideally always check for errors after looping with `rows.Err()`. It's not enough to just `defer rows.Close()`.

Reply:
fixed